src/rt/sched/behaviourcore.h

```diff
@@ -357,6 +357,126 @@ namespace verona::rt
       body->resolve(ec);
     }
 
+    template<TransferOwnership transfer = NoTransfer>
+    static void schedule_many(BehaviourCore** bodies, size_t body_count)
+    {
+      Logging::cout() << "BehaviourCore::schedule_many" << body_count
+                      << Logging::endl;
+
+      size_t count = 0;
+      for (size_t i = 0; i < body_count; i++)
+        count += bodies[i]->count;
+
+      // Execution count - we will remove at least
+      // one from the execution count on finishing phase 2 of the
+      // 2PL. This ensures that the behaviour cannot be
+      // deallocated until we finish phase 2.
+      StackArray<size_t> ec(body_count);
+      for (size_t i = 0; i < body_count; i++)
+        ec[i] = 1;
+
+      // Really want a dynamically sized stack allocation here.
+      StackArray<std::tuple<size_t, Slot*>> indexes(count);
+      size_t idx = 0;
+      for (size_t i = 0; i < body_count; i++)
+      {
+        auto slots = bodies[i]->get_slots();
+        for (size_t j = 0; j < bodies[i]->count; j++)
+        {
+          std::get<0>(indexes[idx]) = i;
+          std::get<1>(indexes[idx]) = &slots[j];
+          idx++;
+        }
+      }
+      auto compare = [](
+                       const std::tuple<size_t, Slot*> i,
+                       const std::tuple<size_t, Slot*> j) {
+#ifdef USE_SYSTEMATIC_TESTING
+        return std::get<1>(i)->cown->id() > std::get<1>(j)->cown->id();
+#else
+        return std::get<1>(i)->cown > std::get<1>(j)->cown;
+#endif
+      };
+
+      if (count > 1)
+        std::sort(indexes.get(), indexes.get() + count, compare);
+
+      // First phase - Acquire phase.
+      Cown *prev_cown = nullptr;
+      for (size_t i = 0; i < count; i++)
+      {
+        auto cown = std::get<1>(indexes[i])->cown;
+        auto body = bodies[std::get<0>(indexes[i])];
+        auto prev = cown->last_slot.exchange(
```

Comment:
I think this needs to account for scheduling multiple behaviours on the same cown.  At the moment you are adding one behaviour at a time.  But that means another scheduler thread could interleave in here.  I think you need to find the set of requests for a particular cown, and enqueue them atomically as a set.  That is

* Chain them together,
* Exchange in the last in the chain,
* Wait for what exchange returned to complete
* Link the first in the chain to what Exchange returns.

Reply:
The release phase would then only need to set_ready on the last in each chain.